runtime/derive/src/lib.rs

```diff
@@ -4,7 +4,103 @@ extern crate proc_macro;
 
 use proc_macro::TokenStream;
 use quote::quote;
-use syn::{Fields, Item, ItemEnum, ItemStruct};
+use syn::{
+    parse::{Parse, ParseStream},
+    Fields, FieldsNamed, Ident, Item, ItemEnum, ItemStruct, Token,
+};
+
+const REQUIRED_IDENT_NAME: &str = "__required__";
+
+struct Args {
+    vars: Vec<ArgsField>,
+}
+
+struct ArgsField {
+    ident: Ident,
+    fields: FieldsNamed,
+}
+
+impl Parse for Args {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut idents = Vec::<ArgsField>::new();
+
+        while input.peek(syn::Ident) {
+            let ident = input.call(Ident::parse)?;
+            idents.push(ArgsField {
+                ident,
+                fields: input.call(FieldsNamed::parse)?,
+            });
+            let _: Option<Token![,]> = input.parse()?;
+        }
+
+        Ok(Args { vars: idents })
+    }
+}
+
+#[derive(Debug)]
+struct TypeParam(syn::Ident);
+
+impl syn::parse::Parse for TypeParam {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let content;
+        syn::parenthesized!(content in input);
+        let typ = content.parse()?;
+        Ok(TypeParam(typ))
+    }
+}
+
+#[derive(Debug)]
+struct TypeParamList(Vec<syn::Ident>);
+
+impl syn::parse::Parse for TypeParamList {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let content;
+        syn::parenthesized!(content in input);
+
+        let mut params: Vec<syn::Ident> = Vec::new();
+
+        while !content.is_empty() {
+            let typ = content.parse()?;
+            params.push(typ);
+
+            if !content.is_empty() {
+                let _comma: syn::Token![,] = content.parse()?;
+            }
+        }
+
+        Ok(TypeParamList(params))
+    }
+}
+
+//intended use is in build.rs with tonic_build's type_attribute(<...>, <...>) to generate type implementation of graph::runtime::ToAscObj
+//attributes: optional list of required fields '__required__{name:TypeName}' and enumerations field decraration with types, i.e. sum{single: ModeInfoSingle,multi: ModeInfoMulti}
```

Comment:
Do we still need support to manually pass required and enumeration explicitly?

Reply:
No, this is meta info derived from proto files